### PR TITLE
Improve directory structure

### PR DESCRIPTION
**Pull request by lm-konda**
## Motivation and Context
As a result of consulting with the two members,
original code's directory structure was not optimal.

So, rewrite partially and add aarch64_neon directory for portability and performance.

This PR is based on #146 

## Description
・remove some directory and make arch directory.
・and separate for each target.
・modify makefile to enable compilation.

The image looks like the following.
```
funcーarchーーgeneric
　　　　　｜＿arm_neon
　　　　　｜＿fpga
　　　　　｜＿aarch64_neon
```

modified:
　modified files in #146 
　dlk/python/dlk/templates/CMakeLists.txt
　dlk/python/dlk/templates/Makefile.tpl
　dlk/python/dlk/templates/cmake/config.cmake
　and other files depend on archtectures

## How has this been tested?
run on AArch64 / ARM / X86 / FPGA

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by lm-konda** [comment]
run dlk test

File: dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp [hunk 1]
```diff
@@ -21,10 +21,10 @@ limitations under the License.
...
 
 #include "global.h"
-#include "func/impl/apply_thresholds.h"
-#include "func/impl/quantized_conv2d_dim2col.h"
-#include "func/impl/quantized_conv2d_tiling.h"
-#include "func/impl/quantized_conv2d_kn2row.h"
+#include "func/apply_thresholds.h"
+#include "func/arch/quantized_conv2d_dim2col.h"
+#include "func/arch/quantized_conv2d_tiling.h"
+#include "func/arch/quantized_conv2d_kn2row.h"
```
Comment (n-nez) [inline dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp]: What rule are you using to decide should header become `func/xxx.h` or `func/arch/xxx.h` ? :thinking: 

Inputitively `arch/<arch name>` level should contain only implementations, while declarations should be somewhere else...

File: dlk/python/dlk/templates/Makefile.tpl [hunk 2]
```diff
@@ -215,8 +225,8 @@ $(LIBS_FPGA): $(LIB_OBJ) $(TVM_OBJ) $(LIB_FPGA_OBJ)
 $(ARS_X86): $(LIB_OBJ) $(TVM_OBJ) $(LIB_X86_OBJ)
 	$(AR) $(LDFLAGS) libdlk_$(NAME).a $(LIB_OBJ) $(TVM_OBJ) $(TVM_X86_LIBS) $(LIB_X86_OBJ)
 
-$(ARS_AARCH64): $(LIB_OBJ) $(TVM_OBJ)
-	$(AR) $(LDFLAGS) libdlk_$(NAME).a $(LIB_OBJ) $(TVM_OBJ) $(TVM_AARCH64_LIBS)
+$(ARS_AARCH64): $(LIB_OBJ) $(TVM_OBJ) $(LIB_AARCH64_OBJ)
+		$(AR) $(LDFLAGS) libdlk_$(NAME).a $(LIB_OBJ) $(TVM_OBJ) $(TVM_AARCH64_LIBS) $(LIB_AARCH64_OBJ)
```
Comment (n-nez) [inline dlk/python/dlk/templates/Makefile.tpl]: could we change indentation to the original one?

File: dlk/python/dlk/templates/Makefile.tpl [hunk 3]
```diff
@@ -194,17 +204,17 @@ $(TARGETS_ARM): $(OBJ) $(TVM_OBJ) $(LIB_ARM_OBJ)
...
 $(TARGETS_X86): $(OBJ) $(TVM_OBJ) $(LIB_X86_OBJ)
 	$(CXX) $(FLAGS) $(OBJ) $(TVM_OBJ) $(LIB_X86_OBJ) -o $@.elf $(CXXFLAGS) $(TVM_X86_LIBS) -pthread -ldl
 
 $(LIBS_X86): $(LIB_OBJ) $(TVM_OBJ) $(LIB_X86_OBJ)
 	$(CXX) $(FLAGS) $(LIB_OBJ) $(TVM_OBJ) $(LIB_X86_OBJ) -o $@.so $(CXXFLAGS) $(TVM_X86_LIBS) -shared -pthread -ldl
 
-$(LIBS_AARCH64): $(LIB_OBJ) $(TVM_OBJ)
-	$(CXX) $(FLAGS) $(LIB_OBJ) $(TVM_OBJ) -o $@.so $(CXXFLAGS) -shared -pthread -ldl
+$(LIBS_AARCH64): $(LIB_OBJ) $(TVM_OBJ) $(LIB_AARCH64_OBJ)
+		 $(CXX) $(FLAGS) $(LIB_OBJ) $(TVM_OBJ) $(LIB_AARCH64_OBJ) -o $@.so $(CXXFLAGS) $(TVM_AARCH64_LIBS)  -shared -pthread -ldl
```
Comment (n-nez) [inline dlk/python/dlk/templates/Makefile.tpl]: could we change indentation to the original one?

File: dlk/python/dlk/templates/Makefile.tpl [hunk 4]
```diff
@@ -194,17 +204,17 @@ $(TARGETS_ARM): $(OBJ) $(TVM_OBJ) $(LIB_ARM_OBJ)
 $(TARGETS_FPGA): $(OBJ) $(TVM_OBJ) $(LIB_FPGA_OBJ)
 	$(CXX) $(FLAGS) $(OBJ) $(TVM_OBJ) $(LIB_FPGA_OBJ) -o $@.elf $(CXXFLAGS) $(TVM_ARM_LIBS) -pthread -ldl
 
-$(TARGETS_AARCH64): $(OBJ) $(TVM_OBJ)
-	$(CXX) $(FLAGS) $(OBJ) $(TVM_OBJ) -o $@.elf $(CXXFLAGS) -pthread -ldl
+$(TARGETS_AARCH64): $(OBJ) $(TVM_OBJ) $(LIB_AARCH64_OBJ)
+		    $(CXX) $(FLAGS) $(OBJ) $(TVM_OBJ) $(LIB_AARCH64_OBJ) -o $@.elf $(CXXFLAGS) $(TVM_AARCH64_OBJ) -pthread -ldl
```
Comment (n-nez) [inline dlk/python/dlk/templates/Makefile.tpl]: could we change indentation to the original one?

File: dlk/python/dlk/templates/Makefile.tpl [hunk 5]
```diff
@@ -138,8 +148,8 @@ clean:
 lm_x86:           CXX = g++
 lm_x86:           FLAGS += $(INCLUDES) -O3 -std=c++0x -g -DUSE_PNG -pthread -g
 
-lm_aarch64:       CXX = aarch64-linux-gnu-g++
-lm_aarch64:       FLAGS += $(INCLUDES) -O3 -std=c++0x -g -DUSE_NEON -DUSE_PNG -pthread -g
+lm_aarch64:           CXX = aarch64-linux-gnu-g++
+lm_aarch64:           FLAGS += $(INCLUDES) -std=c++0x -O3 -DUSE_ASIMD -DUSE_PNG -pthread -g -fopenmp
```
Comment (n-nez) [inline dlk/python/dlk/templates/Makefile.tpl]: could we change indentation to the original one?

File: dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp [hunk 6]
```diff
@@ -62,14 +62,15 @@ void QuantizedConv2D(QUANTIZED_NOT_PACKED input[], T_UINT kernel[],
...
     if ((ic % TilingInTypeBitWidth) == 0) {
-#if defined(USE_NEON) && !defined(RUN_ON_FPGA)
       dlk::impl::QuantizedConv2DTiling(input, kernel, p);
-#else
+    } else {
       dlk::impl::QuantizedConv2DKn2Row(input, kernel, p);
-#endif
+    }
+  } else if (kh == 1 && kw == 1 && padding == 0) {
+    if ((ic % TilingInTypeBitWidth) == 0) {
```
Comment (n-nez) [inline dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp]: Have a strong filling that `QuantizedConv2DKn2Row()` also will not work properly for case `ic % TilingInTypeBitWidth != 0`... :thinking:

File: dlk/python/dlk/templates/Makefile.tpl (hunk 2 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/Makefile.tpl]: OK. I will fix it.

File: dlk/python/dlk/templates/src/func/arch/fpga/quantized_conv2d.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  Measurement::Stop();
+
+  Measurement::Start("QuantizedConv2D_ApplyScalingFactor");
+
+  unsigned out_elems = p.normal_conv_params.output_height *
+                       p.normal_conv_params.output_width *
+                       p.normal_conv_params.output_channels;
+
+  // temporary: (2^n - 1) * (max - min)
+  const T_FLOAT post_qtz_factor = 2.0f / 3.0f;
```
Comment (n-nez) [inline dlk/python/dlk/templates/src/func/arch/fpga/quantized_conv2d.cpp]: @tkclimb, @antonionevado someone knows do we still need hardcoded scaling factors inside of `QuantizedConv2D()` ? Looks very hacky... :scream:

File: dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp (hunk 1 above)
Comment (n-nez) [reply dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp]: Got explanations offline

Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp]: For `/include/func/`, it is a policy to put TensorFlow functions etc., and `/include/include/func/arch/` to put architecture dependent system functions.

File: dlk/python/dlk/templates/Makefile.tpl (hunk 3 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/Makefile.tpl]: OK.

File: dlk/python/dlk/templates/Makefile.tpl (hunk 4 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/Makefile.tpl]: OK.

File: dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp (hunk 6 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/arch/aarch64_neon/quantized_conv2d.cpp]: This is original implementation. Nikolay-san and I are seeing this issue.

File: dlk/python/dlk/templates/Makefile.tpl (hunk 5 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/Makefile.tpl]: OK

**Comment by iizukak** [comment]
@lm-konda Please solve the conflict

**Comment by lm-konda** [comment]
@iizukak All tests have passed!

**Comment by iizukak** [comment]
@lm-konda This PR include AArch64 support.
So, should we merge after AArch64 support ?

**Comment by lm-konda** [comment]
Yes. You are right!

**Comment by iizukak** [comment]
This PR has no progress recently, So I close this PR temporary.

**Pull request closed without merging**